### PR TITLE
journal: limit the journal size in the container

### PR DESCRIPTION
**Pull request by taskset**
The container's /run inherits the physical machine's, and its capacity
s 20% of the physical machine's memory. The memory of the server in the
production environment is usually more than 256G, so /run/log/journal/
may occupy up to 4G (MAX_USE_UPPER). In this way, when dozens or hundreds
of containers are running on the server, OOM may occur.

oci-systemd-hook handles this situation and limits the capacity of /run
and /var/log/journal in the container to 64M
(https://github.com/projectatomic/oci-systemd-hook/blob/master/src/systemdhook.c#L482).
But not all containers can use oci-systemd-hook.

The container environment can already be automatically recognized in systemd,
so the journal size in the container can be limited in systemd to prevent
OOM and/or high disk usage.

### CODE REVIEW

File: src/journal/journal-file.c [hunk 1]
```diff
@@ -3906,6 +3908,7 @@ void journal_default_metrics(JournalMetrics *m, int fd) {
         char a[FORMAT_BYTES_MAX], b[FORMAT_BYTES_MAX], c[FORMAT_BYTES_MAX], d[FORMAT_BYTES_MAX], e[FORMAT_BYTES_MAX];
         struct statvfs ss;
         uint64_t fs_size = 0;
+        int v;
```
Comment (bluca) [inline src/journal/journal-file.c]: style nit: variables should be declared in the inner-most scope - ie, move this inside the "if" block where it's used

Comment (taskset) [reply src/journal/journal-file.c]: OK, thanks.

**Comment by poettering** [comment]
Hmm, what? I don't get it?

Why not make /run smaller?

Or if usage of tmpfs is an issue, why not just make journald write to /var/log/journal, i.e. it's regular behaviour? You basically point it to tmpfs and then complain it uses tmpfs? I don't get it?

**Comment by w-simon** [comment]
> Hmm, what? I don't get it?
> 
> Why not make /run smaller?
> 
> Or if usage of tmpfs is an issue, why not just make journald write to /var/log/journal, i.e. it's regular behaviour? You basically point it to tmpfs and then complain it uses tmpfs? I don't get it?

In the case of errors in writing regular /var/log/journal, the /run/log/journal will also be written.
We found some containers in our production environment, and their /var/log/journal and /run/log/journal both occupy 4G.

Making /run smaller or remounting to a smaller /var/log/journal is also a way, but they have a larger impact.

Thanks.

**Comment by poettering** [comment]
I must say, I disagree with such special casing. We can't guess how people use their containers. I run some containers with beefy limits myself (compile stuff), and I don't think they should get their journals reduced in size just like that.

I think the best approach is to make everything scaled off the resources the container itself gets passed in, like we already do: i.e. I think it's more interesting to say "give this container x gb space in /run", rather than to say "give this container x gb journal space in /run": we should always scale down from what the immediate amount the surrounding system has, and not second guess this. If you define a beefy container we should use a bigger journal, and if you define a lean container with little space in /run, we should use a smaller journal.

Moreover, this is a knob I think that should be dependent on parameters the outside contianer manager controls, and our own code should only work on top of that. Now the knob to control /run usage is the size the container manager picks for it.

I hope this makes some sense?

Or to say this differently: if you lower your /run size for the container the upper limit of the runtime journal shrinks anyway already, and that's how it should be. Hence closing.

**Pull request closed without merging**